### PR TITLE
Pull request for gslab_python#29: Implements paths solution

### PR DESCRIPTION
**Pull request by lboxell**

### CODE REVIEW

**Comment by lboxell** [comment]
@yuchuan2016, Can you PR this? The original task is https://github.com/gslab-econ/gslab_python/issues/29.

**Comment by yuchuan2016** [comment]
@lboxell , this looks good to me! One thing is that should we check prerequisites added (python module, R package and Stata ado file) in the `setup.py`?

**Comment by lboxell** [comment]
@yuchuan2016,

Good point. I've made changes to check the installation of these files. If you want to confirm things still work on windows, that would be great.

Let me know if there is anything else.

**Comment by yuchuan2016** [comment]
Hi @lboxell ,

I'm having some difficulty in specifying the correct Stata flavor on Windows. When I leave it as blank (`stata_flavor: ""`), the `command` from `check_stata(sf)` will be `'%%STATAEXE%% /e do %s '` Then when running
```
subprocess.check_output(command % '"which %s"' % pkg, shell = True)
```
I get an error
```
ValueError: unsupported format character 'S' (0x53) at index 1
```
I think it's due to the `%%S` in the command.

Another thing is about backslash. I got the error
```
scons: *** [build\data\data.txt] build\data/sconscript.log: No such file or directory
```
But if I remove `, CL_ARG = target_file_name`, this error will disappear.

**Comment by lboxell** [comment]
@yuchuan2016,

Good catch. I think I fixed the stata error with the commit above. 

For the backslash error, is that something I introduced in this task or do you think that is a pre-existing gslab_python bug?

**Comment by yuchuan2016** [comment]
Hi @lboxell ,

I think the command needs to changed to `%STATAEXE% /e which yaml` for it to work on Window. Now it's `%%STATAEXE%% /e do "which yaml"`

About the other error, I think it was introduced by the command line argument. When running `source/data/SConscript`, a file named `data'` was produced at the root with content
>Fatal error: cannot open file ''--args': No such file or directory

I think it was introduced from `build_r.py`
image

**Comment by lboxell** [comment]
@arosenbe

File: setup.py [hunk 1]
```diff
@@ -91,10 +83,54 @@ def check_gitlfs():
     from gslab_scons.misc import check_lfs
     check_lfs()
 
+def check_yamls():
+    if not os.path.isfile("constants.yaml"):
+        raise PrerequisiteError("constants.yaml file does not exist. Please create it.")
+    if not os.path.isfile("user-config.yaml"):
+        raise PrerequisiteError("user-config.yaml file does not exist. Please create it.")
```
Comment (arosenbe) [inline setup.py]: Should this be handled in a loop?

File: SConstruct [hunk 2]
```diff
@@ -2,20 +2,16 @@
...
-mode      = ARGUMENTS.get('mode', 'develop') # Gets mode; defaults to 'develop'
-vers      = ARGUMENTS.get('version', '') # Gets release version; defaults to ''
-sf        = ARGUMENTS.get('sf', None) # Gets user supplied stata or defaults to None
-cache_dir = '/Users/%s/Google Drive/cache/large_template'    % os.environ['USER']
 
-# Test for proper prerequisites and setup; import gslab_scons after passing setup
+# Test for proper prerequisites and setup
 from setup import setup_test
-setup_test(mode, vers, sf, cache_dir)
+[user_configs, mode, sf, cache_dir] = setup_test(ARGUMENTS)
```
Comment (lboxell) [inline SConstruct]: Does 'setup' and 'test'?

File: SConstruct [hunk 3]
```diff
@@ -30,6 +26,10 @@ env.Decider('MD5-timestamp') # Only computes hash if time-stamp changed
 env.EXTENSIONS = ['.eps', '.pdf', '.lyx'] # Extensions to be used when scanning for source files in BuildLyx.
 SourceFileScanner.add_scanner('.lyx', Scanner(gslab_scons.misc.lyx_scan, recursive = True))
 
+# Load paths
+env['PATHS'] = yaml.load(open("constants.yaml", 'rU'))
```
Comment (lboxell) [inline SConstruct]: I don't think this in general is standard scons practice, so I wouldn't expect there to be an 'accepted scons-specific paths variable name'.

File: constants.yaml [hunk 4]
```diff
@@ -0,0 +1,17 @@
+# Build Directories
```
Comment (lboxell) [inline constants.yaml]: As its currently set up, correct. This is easy to change for any given project though.

File: README.md [hunk 5]
```diff
@@ -5,8 +5,11 @@ Using the repository template
```
Comment (arosenbe) [inline README.md]: I think the readme could use some sprucing up. perhaps that's another task where we improve all our docs.

File: setup.py [hunk 6]
```diff
@@ -1,77 +1,69 @@
...
-    check_gslab_tools()
-    check_stata(sf)
     check_r()
     check_lyx()
     check_metropolis()
     check_gitlfs()
+    check_yamls()
+
+    # Loads arguments and configurations
+    import yaml
```
Comment (lboxell) [inline setup.py]: Yes. But, its not in scope anymore.
```
Python 2.7.13 |Anaconda 4.3.0 (x86_64)| (default, Dec 20 2016, 23:05:08) 
[GCC 4.2.1 Compatible Apple LLVM 6.0 (clang-600.0.57)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
Anaconda is brought to you by Continuum Analytics.
Please check out: http://continuum.io/thanks and https://anaconda.org
>>> def test():
...     import sys
... 
>>> test()
>>> sys.platform
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
NameError: name 'sys' is not defined
>>> 
```

File: source/data/build.R [hunk 7]
```diff
@@ -1,7 +1,10 @@
-args = commandArgs(TRUE)
+library(yaml)
+PATHS <- yaml.load_file('constants.yaml')
+args  <- commandArgs(TRUE)
+
 main <- function(i){
     x    <- seq(-10, 10, 0.1)
-    write.table(x, sprintf("build/data/%s.txt", i), sep = "|", 
+    write.table(x, sprintf("%s/%s.txt", PATHS$build$data, i), sep = "|",
```
Comment (arosenbe) [inline source/data/build.R]: We seem to be having some trouble with floating point precision for numbers less than one with this command. Probably not what we want to be demonstrating.

File: source/tables/SConscript [hunk 8]
```diff
@@ -1,5 +1,5 @@
 Import('env')
 
-env.Tablefill(target = ['#build/tables/table.lyx'],
-              source = ['#source/tables/table.lyx',
-                        '#build/analysis/table.txt'])
\ No newline at end of file
+env.Tablefill(target = ['#%s/table.lyx' % env['PATHS']['build']['tables']],
+              source = ['#%s/table.lyx' % env['PATHS']['source']['tables'],
+                        '#%s/table.txt' % env['PATHS']['build']['analysis']])
```
Comment (arosenbe) [inline source/tables/SConscript]: Can we add a carriage return at the end of this file.

File: SConstruct (hunk 3 above)
Comment (lboxell) [inline SConstruct]: Though I guess it could be confusing with the standard os `PATH` variable.

File: source/analysis/analysis.do [hunk 9]
```diff
@@ -2,17 +2,23 @@ version 14
 clear all
 set more off
 
+
 program main
-    import delimited "build/data/data.txt", delimiters("|") varnames(1) clear
+    yaml read YAML using constants.yaml
```
Comment (arosenbe) [inline source/analysis/analysis.do]: Why do we prefer to read the yaml in stata within the main function but before the main function in R?

**Comment by lboxell** [comment]
@arosenbe,

I implemented the comments I thought should be implemented. Chuan already did the PR for this task and I think we just need to try and fix what is going on in this series of comments regarding windows: https://github.com/gslab-econ/template/pull/35#issuecomment-277168498

File: source/analysis/analysis.do (hunk 9 above)
Comment (lboxell) [inline source/analysis/analysis.do]: It's a scope issue. It didn't seem to work when I had the stata yaml calls outside of the main function. In politext, I think we normally called the paths variable in the same place where we had our `library()` calls which is what I defaulted to here. 
```
Log created:    2017-02-06 17:09:18
Log completed:  2017-02-06 17:09:19
 

  ___  ____  ____  ____  ____ (R)
 /__    /   ____/   /   ____/
___/   /   /___/   /   /___/   14.2   Copyright 1985-2015 StataCorp LP
  Statistics/Data Analysis            StataCorp
                                      4905 Lakeway Drive
     MP - Parallel Edition            College Station, Texas 77845 USA
                                      800-STATA-PC        http://www.stata.com
                                      979-696-4600        stata@stata.com
                                      979-696-4601 (fax)

Unlimited-user 4-core Stata network license expires 21 Jul 2017:
       Serial number:  501409012939
         Licensed to:  Economics
                       Stanford University

Notes:
      1.  Stata is running in batch mode.
      2.  Unicode is supported; see help unicode_advice.
      3.  More than 2 billion observations are allowed; see help obs_advice.
      4.  Maximum number of variables is set to 5000; see help set_maxvar.


running /Applications/Stata/profile.do ...

. do "source/analysis/analysis.do" 

. version 14

. clear all

. set more off

. 
. yaml read YAML using constants.yaml

. yaml local build_data=YAML.build.data

. yaml local build_analysis=YAML.build.analysis

. 
. program main
  1. 
.     import delimited "`build_data'/data.txt", delimiters("|") varnames(1) clear
  2.     hist x, discrete width(0.5)
  3.     graph export "`build_analysis'/plot.eps", replace
  4. 
.     sum x
  5.     file open  outfile using "`build_analysis'/table.txt", write replace
  6.     file write outfile "<tab:table>" _n
  7.     file write outfile (r(mean)) _n (r(sd)) _n (r(max)) _n (r(min))
  8.     file close outfile
  9. end

. 
. 
. * EXECUTE
. main
file /data.txt not found
r(601);

end of do-file
r(601);
```

File: README.md (hunk 5 above)
Comment (lboxell) [inline README.md]: Probably outside the scope of this task.

File: source/tables/SConscript (hunk 8 above)
Comment (lboxell) [inline source/tables/SConscript]: Done.

File: source/data/build.R (hunk 7 above)
Comment (lboxell) [inline source/data/build.R]: Can probably make a note with the previous comment on the outline to rethink this later if we want.

**Comment by lboxell** [comment]
Opened https://github.com/gslab-econ/gslab_python/issues/57 to address the windows issues. Going to squash and merge.

**Pull request merged**